daemon/cmd/watchdogs.go

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/pflag"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/datapath/loader"
+	"github.com/cilium/cilium/pkg/endpoint"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/promise"
+)
+
+const epBPFProgWatchdog = "ep-bpf-prog-watchdog"
+
+type epBPFProgWatchdogConfig struct {
+	EndpointBPFProgWatchdogRate    time.Duration
+	EndpointBPFProgWatchdogEnabled bool
+}
+
+func (c epBPFProgWatchdogConfig) Flags(flags *pflag.FlagSet) {
+	flags.Bool("endpoint-bpf-prog-watchdog-enabled", c.EndpointBPFProgWatchdogEnabled,
+		"Enable endpoint BPF program watchdog")
+
+	flags.Duration("endpoint-bpf-prog-watchdog-rate", c.EndpointBPFProgWatchdogRate,
+		"Duration to trigger endpoint BPF programs load check watchdog")
+}
+
+type epBPFProgWatchdogParams struct {
+	cell.In
+
+	Config        epBPFProgWatchdogConfig
+	Logger        logrus.FieldLogger
+	Lifecycle     hive.Lifecycle
+	DaemonPromise promise.Promise[*Daemon]
+}
+
+var (
+	// EndpointBPFrogWatchdogCell triggers a job to ensure device tc programs remain loaded.
+	EndpointBPFrogWatchdogCell = cell.Module(
+		epBPFProgWatchdog,
+		"Periodically checks that endpoints BPF programs remain loaded",
+
+		cell.Config(epBPFProgWatchdogConfigDefault),
+		cell.Invoke(registerEndpointBPFProgWatchdog),
+	)
+
+	epBPFProgWatchdogConfigDefault = epBPFProgWatchdogConfig{
+		EndpointBPFProgWatchdogEnabled: true,
+		EndpointBPFProgWatchdogRate:    30 * time.Second,
+	}
+)
+
+func registerEndpointBPFProgWatchdog(p epBPFProgWatchdogParams) {
+	if !p.Config.EndpointBPFProgWatchdogEnabled {
+		return
+	}
+
+	var (
+		ctx, cancel = context.WithCancel(context.Background())
+		mgr         = controller.NewManager()
+	)
+	p.Lifecycle.Append(hive.Hook{
+		OnStart: func(hive.HookContext) error {
+			mgr.UpdateController(
+				epBPFProgWatchdog,
+				controller.ControllerParams{
+					Group: controller.NewGroup(epBPFProgWatchdog),
+					DoFunc: func(ctx context.Context) error {
+						d, err := p.DaemonPromise.Await(context.Background())
+						if err != nil {
+							return err
+						}
+						if d.ctx.Err() != nil {
+							mgr.RemoveAllAndWait()
```

Comment:
I don't think we want to do this because it will kill all of the controllers. This controller should only care about itself, otherwise we'll have this code and the daemon-stopping logic stepping on each other's toes.

Reply:
This should only impact this cell's controller and not the daemon's.

Reply:
@derailed is right, it's only affecting the manager that's created here.

But do not call `RemoveAllAndWait` from `DoFunc`. It should only be called from the stop hook.
And do not use `d.ctx`. This cell is separate from the daemon and thus should not use its context. And since this is depending on the daemon, it means it'll be stopped before daemon and thus `d.ctx` will never be cancelled before the stop hook of this cell is called, e.g. this path is never reached.

Reply:
Tx Jussi! Good point.